### DISCUSSION TITLE
Custom persistence provider implementation guide with Elsa 3.0

### DISCUSSION THREAD

**Discussion by vnmnm** [elsa-workflows/elsa-core#4090] [Q&A]
Hi Team,

With Elsa 2.0, we followed this guide at [link](https://elsa-workflows.github.io/elsa-core/docs/extensibility/extensibility-persistence) and also referred existing implementations to implement our own provider. It was clear as to what all interfaces we needed to implement in Elsa 2.0.

However, when I refer the Elsa 3.0 source code, the interfaces are bit different and it's not clear how to implement a custom provider. This [link](https://v3.elsaworkflows.io/docs/installation/aspnet-apps-workflow-server#custom-persistence) in the user guide points to a non-existent URL and throws 404 error.

It would be great if I can get some guidance on implementing custom persistence provider with Elsa 3.0.

**Comment by sfmskywalker** [comment]
I'll work on the documentation for that 👍🏻

**Reply by sfmskywalker** [reply]
Some basic pointers have been provided: https://v3.elsaworkflows.io/docs/extensibility/custom-persistence
Please let me know if this is helpful or not, so we can continue adding details as necessary.

**Reply by vnmnm** [reply]
@sfmskywalker Thank you for adding this documentation. This is definitely helpful. Also, abstractions like `IWorkflowDefinitionService` etc that depend on `IWorkflowDefinitionStore` and other stores are all registered as singleton. This would cause issues when we want to implement database per tenant kind of multi-tenancy in which case we would need all of these to be scoped. Not sure if I am missing something w.r.t singleton regsitrations.

**Reply by sfmskywalker** [reply]
Sorry, I forgot to answer your question about the Singleton lifetime.

The reason I use Singleton is because it means the instances are created just once, and can then be reused throughout the lifetime of the application. I.e., it's a nice performance improvement (in theory, at least) we get for free.

The current persistence implementations can be singleton because they perform very small units of work, like within the method itself.
For example, the EF Core implementations inject an `IDbContextFactory` (a singleton), from which the methods create a `DbContext`, which gets disposed immediately before exiting the method.

If your custom persistence provider depends on scoped services, right now the only way to handle this is to inject an `IServiceScopeFactory` and then use it to create a child scope. Then you can resolve scoped services from this scope.

Example (pseudo code to keep it short):

```csharp
class CustomWorkflowDefinitionStore : IWorkflowDefinitionStore
{
   ctor(private IServiceScopeFactory serviceScopeFactory)
   {
   }
   
   public async Task<WorkflowDefinition?> FindAsync(WorkflowDefinitionFilter filter)
   {
      using var scope = _serviceScopeFactory.CreateScope();
      var myDbAccessService = scope.ServiceProvider.GetRequiredService<IMyScopedDbAccessService>();
       ...
   }
}
```

That being said, we may yet need to change the lifetime scope of persistence (and other) services when we work on multi-tenancy, and the easiest way might be to use a scoped lifetime in the end.
